293 parser tutorial with exercise (#294)
FAIRmat-NFDI/nomad-docs: docs/tutorial/develop_plugin/plugin_structure.md

2. Generate a plugin project using the official NOMAD cookiecutter template
3. Set up a development and testing environment for the plugin

In the following tutorials, you will create a custom schema package and a parser.

---

## Before you begin

This tutorial assumes basic familiarity with Python and Git and is intended for users who want to extend NOMAD with custom functionality.

Before starting this tutorial, make sure you have the following:

1. **GitHub account**  
   Required to create and manage the plugin repository. You can create a free account at [github.com/signup](https://github.com/signup){:target="_blank" rel="noopener"}.

2. **Basic understanding of Python**  
   You should be comfortable reading and writing basic Python code, including modules, functions,
   and classes.

3. **Basic understanding of NOMAD metainfo**  
   Familiarity with NOMAD’s metainfo system is helpful. If needed, review [FAIRmat Tutorial 8](https://www.fairmat-nfdi.eu/events/fairmat-tutorial-8/tutorial-8-materials){:target="_blank" rel="noopener"}.

4. **Local or cloud-based development environment**  
   You need either:

>     - A Linux-based local machine with Python ≥ 3.12, git, and any Integrated Development Environment (IDE), or

Reviewer: This is not really true anymore is it? This tutorial should run fine on Windows as well.

```suggestion
    - A local machine with Python ≥ 3.12, git, and any Integrated Development Environment (IDE), or
```

    - [what is cruft](https://cruft.github.io/cruft/){:target="_blank" rel="noopener"}

---

## Create a plugin repository

First, you will version-control your NOMAD plugin by creating a GitHub repository from the official template. Start from the official GitHub template repository at
[github.com/FAIRmat-NFDI/nomad-plugin-template](https://github.com/FAIRmat-NFDI/nomad-plugin-template){:target="_blank" rel="noopener"}.

To create a new repository from the template, select **Use this template** and then choose
**Create a new repository**. You must be logged in to GitHub to see this option.

![Use template](../images/use_template_dark.png#gh-dark-mode-only)
![Use template](../images/use_template_light.png#gh-light-mode-only)

Enter a repository name (for example, `nomad-optical-microscopy`) and select **Create repository** to complete the setup.
<!-- TODO: add image slider to show the two steps -->

## Generate the plugin structure

Next, you will generate the initial structure of the plugin by applying the official NOMAD cookiecutter template.

### Choose a development environment

You can proceed in one of two ways:

1. Use GitHub Codespaces (cloud-based development), or

> 2. Develop locally on Linux.

Reviewer: Again
```suggestion
2. Develop locally.
```

**Using GitHub codespaces**

To use a GitHub codespace for the plugin development, click on the **<> Code** button in the repository and choose **Create codespace on main**.

![Use codepace](../images/codespace_dark.png#gh-dark-mode-only)
![Use codespace](../images/codespace_light.png#gh-light-mode-only)

**Developing locally**

> If you prefer to work locally on a Linux machine, click on the **<> Code** button in the repository and choose the **“Local”** tab, copy the repository URL, and clone it to a selected location in your machine by running in terminal:

Reviewer: Again
```suggestion
If you prefer to work locally, click on the **<> Code** button in the repository and choose the **“Local”** tab, copy the repository URL, and clone it to a selected location in your machine by running in terminal:
```

!!! success "You have just created a minimal NOMAD plugin with a plugin entry point for a schema package"
    ```no-highlight
    optical-microscopy/
    ├── LICENSE
    ├── MANIFEST.in
    ├── README.md
    ├── docs
    │   └── ...
    ├── mkdocs.yml
    ├── move_template_files.sh
    ├── pyproject.toml
    ├── src
    │   └── optical_microscopy
    │       ├── __init__.py
    │       ├── parsers
    │       │   ├── __init__.py
    │       │   └── parser.py
    │       └── schema_packages
    │           ├── __init__.py
    │           └── schema_package.py
    └── tests
        ├── conftest.py
        ├── data
        │   └── test.archive.yaml
        ├── parsers
        │   └── test_parser.py
        └── schema_packages
            └── test_schema.py
    ```

The plugin is generated in a subdirectory. Move the files to the repository root using the provided helper script:

```sh
sh optical-microscopy/move_template_files.sh
```

Finally, add the files to Git and commit the changes you have made:

```sh
git add -A
git commit -m "Generated plugin from cookiecutter template"
git push
```

> **Enable cruft updates**

Reviewer: I'm not sure if we still recommend people to do this but we can look at it after the tutorial.

![Use template](../images/workflow_permissions_dark.png#gh-dark-mode-only)
![Use template](../images/workflow_permissions_light.png#gh-light-mode-only)

## Developing the plugin

The structure of the plugin is now ready for development for your specific purposes. If you plan to work with a single plugin and can avoid using NOMAD GUI functionality, stand-alone installation of the plugin will be sufficient. If you plan to work with multiple plugins or GUI-specific functionality, or if you wish to develop the core NOMAD package, we recommend using a dedicated [`nomad-distro-dev`](https://github.com/FAIRmat-NFDI/nomad-distro-dev){:target="_blank" rel="noopener"} development environment.

### Option 1: Stand-alone installation of the plugin

In this step, you will set up a Python environment and install the plugin for local development. This can be done conveniently in one step using [uv](https://docs.astral.sh/uv/getting-started/installation/){:target="_blank" rel="noopener"} or in several steps with pip.

**(Recommended) Installation with uv**

Open the terminal, navigate to the folder with your plugin using `cd`. Set up the plugin:

```sh
uv sync --extra dev
```

> After that many commands can be run in the newly created Python environment by using:

Reviewer: This sounds a bit strange, I would change it to something like:

```suggestion
This sets up a dynamic Python environment. To run a python command, use:
```

```sh
uv run <command>
```

**(Alternative) Installation with pip**

Open the terminal, navigate to the folder with your plugin using `cd`. Create a virtual environment using Python 3.12 and activate it:

```sh
python3.12 -m venv .pyenv
source .pyenv/bin/activate
```

Install the plugin package in editable mode using the NOMAD package registry:

```sh
pip install --upgrade pip
pip install -e '.[dev]'
```

### Option 2: `nomad-distro-dev`

> This option should be used with a local Linux-based machine.

Reviewer: Again, nomad-distro-dev should also work on Windows.

Start with forking [`nomad-distro-dev`](https://github.com/FAIRmat-NFDI/nomad-distro-dev){:target="_blank" rel="noopener"} repository (`Fork` -> `Create a new fork` in the upper right part of the page). You will also need the following additional software installed on your system:

- [Docker](https://docs.docker.com/engine/install/){:target="_blank" rel="noopener"} - generally, only `docker-compose` functionality will be needed

- [uv](https://docs.astral.sh/uv/getting-started/installation/){:target="_blank" rel="noopener"} python package manager, version 0.5.14 or above

- [node.js](https://nodejs.org/en){:target="_blank" rel="noopener"} version 20 or above and [yarn](https://classic.yarnpkg.com/en/docs/install){:target="_blank" rel="noopener"} version 1.22 or above are necessary to run the GUI

> Then, check `Basic infra` and follow the `Step-by-Step Setup` instructions in the `nomad-distro-dev` readme file.

Reviewer: ```suggestion
Then, follow the instructions in the `nomad-distro-dev` readme file under the `Basic infra` and `Step-by-Step Setup` headings.
```

## Next steps

> Specific details for developing schema packages and parsers are provided in corresponding tutorials.

Reviewer: Sounds a bit overly formal...
```suggestion
In the next tutorials we will learn how to add data schemas and file parsers to our newly created plugin.
```

Commit all your changes in the plugin to git:

```sh
git add -A
git commit -m "Short description of the changes"
git push

> ```

Reviewer: But now there are now changes here? I would skip this and have it in the tutorials when someone makes changes.
```suggestion
```
